File: tests/generators/test_data_generator.py
"""
Test data generation utilities for Creatio AI Knowledge Hub
"""
import json
import random
import sqlite3
import tempfile
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Any, Optional
from faker import Faker
from factory import Factory, Sequence, LazyAttribute, fuzzy

fake = Faker()


class VideoContentFactory(Factory):
    """Factory for generating video content test data"""
    
    id = Sequence(lambda n: f"video-{n:04d}")
    title = LazyAttribute(lambda obj: fake.sentence(nb_words=4)[:-1])
    file_path = LazyAttribute(lambda obj: f"/videos/{obj.id}.mp4")
    duration = fuzzy.FuzzyFloat(30.0, 1800.0)  # 30 seconds to 30 minutes
    transcript = LazyAttribute(lambda obj: fake.text(max_nb_chars=2000))
    summary = LazyAttribute(lambda obj: fake.paragraph(nb_sentences=3))
    complexity_level = fuzzy.FuzzyChoice(['beginner', 'intermediate', 'advanced'])
    created_at = fuzzy.FuzzyDateTime(
        start_dt=datetime.now() - timedelta(days=365),
        end_dt=datetime.now()
    )
    
    @LazyAttribute
    def topics(obj):
        topics = ['creatio', 'bpm', 'crm', 'development', 'javascript', 'c#', 'configuration']
        return random.sample(topics, k=random.randint(2, 5))
    
    @LazyAttribute
    def commands(obj):
        commands = [
            'CreateSection', 'AddField', 'ConfigureProcess', 'SetupIntegration',
            'CreateEntity', 'ModifySchema', 'AddBusinessRule', 'CreateLookup'
        ]
        return random.sample(commands, k=random.randint(1, 4))
    
    @LazyAttribute
    def api_references(obj):
        apis = [
            'EntitySchema', 'UserConnection', 'ProcessSchema', 'ServiceApi',
            'ConfigurationService', 'MessagePublisher', 'EntitySchemaManager'
        ]
        return random.sample(apis, k=random.randint(1, 3))
    
    @LazyAttribute
    def code_examples(obj):
        examples = []
        languages = ['javascript', 'csharp', 'sql']
        for lang in random.sample(languages, k=random.randint(1, 2)):
            if lang == 'javascript':
                code = "var schema = this.Ext.create('EntitySchema', { name: 'Contact' });"
            elif lang == 'csharp':
                code = "var entity = UserConnection.EntitySchemaManager.GetInstanceByName('Contact');"
            else:
                code = "SELECT Id, Name FROM Contact WHERE CreatedOn > @StartDate"
            
            examples.append({
                'language': lang,
                'code': code,
                'description': fake.sentence()
            })
        return examples


class PDFContentFactory(Factory):
    """Factory for generating PDF content test data"""
    
    id = Sequence(lambda n: f"pdf-{n:04d}")
    title = LazyAttribute(lambda obj: fake.sentence(nb_words=6)[:-1])
    file_path = LazyAttribute(lambda obj: f"/pdfs/{obj.id}.pdf")
    page_count = fuzzy.FuzzyInteger(5, 50)
    content = LazyAttribute(lambda obj: fake.text(max_nb_chars=5000))
    created_at = fuzzy.FuzzyDateTime(
        start_dt=datetime.now() - timedelta(days=365),
        end_dt=datetime.now()
    )
    
    @LazyAttribute
    def sections(obj):
        sections = []
        section_count = random.randint(3, 8)
        for i in range(section_count):
            sections.append({
                'title': fake.sentence(nb_words=4)[:-1],
                'content': fake.paragraph(nb_sentences=5),
                'page_number': random.randint(1, obj.page_count)
            })
        return sections
    
    @LazyAttribute
    def topics(obj):
        topics = ['creatio', 'documentation', 'api', 'configuration', 'development', 'architecture']
        return random.sample(topics, k=random.randint(2, 4))


class CommandFactory(Factory):
    """Factory for generating command test data"""
    
    id = Sequence(lambda n: n)
    command = LazyAttribute(lambda obj: fake.word().title() + random.choice(['Section', 'Field', 'Process', 'Rule']))
    description = LazyAttribute(lambda obj: fake.sentence())
    category = fuzzy.FuzzyChoice(['UI', 'Business Logic', 'Integration', 'Configuration', 'Data'])
    parameters = LazyAttribute(lambda obj: [fake.word() for _ in range(random.randint(0, 3))])


class TestDataGenerator:
    """Main test data generator class"""
    
    def __init__(self, db_path: Optional[str] = None):
        self.db_path = db_path or tempfile.mktemp(suffix='.db')
        self.conn = None
    
    def setup_database(self):
        """Setup test database with schema"""
        self.conn = sqlite3.connect(self.db_path)
        cursor = self.conn.cursor()
        
        # Create videos table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS videos (
                id TEXT PRIMARY KEY,
                title TEXT NOT NULL,
                file_path TEXT NOT NULL,
                duration REAL,
                transcript TEXT,
                summary TEXT,
                topics TEXT,
                complexity_level TEXT,
                commands TEXT,
                api_references TEXT,
                code_examples TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        # Create pdfs table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS pdfs (
                id TEXT PRIMARY KEY,
                title TEXT NOT NULL,
                file_path TEXT NOT NULL,
                page_count INTEGER,
                content TEXT,
                sections TEXT,
                topics TEXT,
                commands TEXT,
                api_references TEXT,
                code_examples TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        # Create commands table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS commands (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                command TEXT NOT NULL,
                description TEXT,
                category TEXT,
                parameters TEXT
            )
        ''')
        
        self.conn.commit()
    
    def generate_videos(self, count: int = 50) -> List[Dict[str, Any]]:
        """Generate video test data"""
        videos = []
        for _ in range(count):
            video_data = VideoContentFactory()
            videos.append({
                'id': video_data.id,
                'title': video_data.title,
                'file_path': video_data.file_path,
                'duration': video_data.duration,
                'transcript': video_data.transcript,
                'summary': video_data.summary,
                'topics': json.dumps(video_data.topics),
                'complexity_level': video_data.complexity_level,
                'commands': json.dumps(video_data.commands),
                'api_references': json.dumps(video_data.api_references),
                'code_examples': json.dumps(video_data.code_examples),
                'created_at': video_data.created_at.isoformat()
            })
        return videos
    
    def generate_pdfs(self, count: int = 30) -> List[Dict[str, Any]]:
        """Generate PDF test data"""
        pdfs = []
        for _ in range(count):
            pdf_data = PDFContentFactory()
            pdfs.append({
                'id': pdf_data.id,
                'title': pdf_data.title,
                'file_path': pdf_data.file_path,
                'page_count': pdf_data.page_count,
                'content': pdf_data.content,
                'sections': json.dumps(pdf_data.sections),
                'topics': json.dumps(pdf_data.topics),
                'commands': json.dumps([]),
                'api_references': json.dumps([]),
                'code_examples': json.dumps([]),
                'created_at': pdf_data.created_at.isoformat()
            })
        return pdfs
    
    def generate_commands(self, count: int = 20) -> List[Dict[str, Any]]:
        """Generate command test data"""
        commands = []
        for _ in range(count):
            command_data = CommandFactory()
            commands.append({
                'command': command_data.command,
                'description': command_data.description,
                'category': command_data.category,
                'parameters': json.dumps(command_data.parameters)
            })
        return commands
    
    def populate_database(self, video_count: int = 50, pdf_count: int = 30, command_count: int = 20):
        """Populate database with test data"""
        if not self.conn:
            self.setup_database()
        
        cursor = self.conn.cursor()
        
        # Insert videos
        videos = self.generate_videos(video_count)
        for video in videos:
            cursor.execute('''
                INSERT INTO videos (id, title, file_path, duration, transcript, summary, 
                                  topics, complexity_level, commands, api_references, 
                                  code_examples, created_at)
                VALUES (:id, :title, :file_path, :duration, :transcript, :summary,
                       :topics, :complexity_level, :commands, :api_references,
                       :code_examples, :created_at)
            ''', video)
        
        # Insert PDFs
        pdfs = self.generate_pdfs(pdf_count)
        for pdf in pdfs:
            cursor.execute('''
                INSERT INTO pdfs (id, title, file_path, page_count, content, sections,
                                topics, commands, api_references, code_examples, created_at)
                VALUES (:id, :title, :file_path, :page_count, :content, :sections,
                       :topics, :commands, :api_references, :code_examples, :created_at)
            ''', pdf)
        
        # Insert commands
        commands = self.generate_commands(command_count)
        for command in commands:
            cursor.execute('''
                INSERT INTO commands (command, description, category, parameters)
                VALUES (:command, :description, :category, :parameters)
            ''', command)
        
        self.conn.commit()
    
    def export_to_json(self, output_path: str):
        """Export test data to JSON file"""
        if not self.conn:
            raise ValueError("Database not setup")
        
        cursor = self.conn.cursor()
        
        # Get all data
        cursor.execute('SELECT * FROM videos')
        videos = [dict(zip([col[0] for col in cursor.description], row)) for row in cursor.fetchall()]
        
        cursor.execute('SELECT * FROM pdfs')
        pdfs = [dict(zip([col[0] for col in cursor.description], row)) for row in cursor.fetchall()]
        
        cursor.execute('SELECT * FROM commands')
        commands = [dict(zip([col[0] for col in cursor.description], row)) for row in cursor.fetchall()]
        
        test_data = {
            'videos': videos,
            'pdfs': pdfs,
            'commands': commands,
            'metadata': {
                'generated_at': datetime.now().isoformat(),
                'video_count': len(videos),
                'pdf_count': len(pdfs),
                'command_count': len(commands)
            }
        }
        
        with open(output_path, 'w') as f:
            json.dump(test_data, f, indent=2, default=str)
    
    def close(self):
        """Close database connection"""
        if self.conn:
            self.conn.close()


def main():
    """Generate test data"""
    generator = TestDataGenerator()
    generator.setup_database()
    generator.populate_database(video_count=100, pdf_count=50, command_count=30)
    
    # Export to JSON
    output_path = Path(__file__).parent / 'test_data.json'
    generator.export_to_json(str(output_path))
    
    print(f"Test data generated and saved to {output_path}")
    print(f"Database saved to {generator.db_path}")
    
    generator.close()


if __name__ == '__main__':
    main()

File: tests/generators/test-data-generator.ts
/**
 * Test data generation utilities for JavaScript/TypeScript tests
 */
import { faker } from '@faker-js/faker';
import fs from 'fs/promises';
import path from 'path';

export interface VideoContent {
  id: string;
  title: string;
  filePath: string;
  duration: number;
  transcript: string;
  summary: string;
  topics: string[];
  complexityLevel: 'beginner' | 'intermediate' | 'advanced';
  commands: string[];
  apiReferences: string[];
  codeExamples: CodeExample[];
  createdAt: Date;
}

export interface PDFContent {
  id: string;
  title: string;
  filePath: string;
  pageCount: number;
  content: string;
  sections: PDFSection[];
  topics: string[];
  commands: string[];
  apiReferences: string[];
  codeExamples: CodeExample[];
  createdAt: Date;
}

export interface CodeExample {
  language: string;
  code: string;
  description: string;
}

export interface PDFSection {
  title: string;
  content: string;
  pageNumber: number;
}

export interface Command {
  id: number;
  command: string;
  description: string;
  category: string;
  parameters: string[];
}

export class TestDataGenerator {
  private static readonly TOPICS = [
    'creatio', 'bpm', 'crm', 'development', 'javascript', 'c#', 'configuration',
    'api', 'documentation', 'architecture', 'integration', 'workflow'
  ];

  private static readonly COMMANDS = [
    'CreateSection', 'AddField', 'ConfigureProcess', 'SetupIntegration',
    'CreateEntity', 'ModifySchema', 'AddBusinessRule', 'CreateLookup',
    'DefinePermissions', 'ConfigureUI', 'SetupValidation', 'CreateReport'
  ];

  private static readonly API_REFERENCES = [
    'EntitySchema', 'UserConnection', 'ProcessSchema', 'ServiceApi',
    'ConfigurationService', 'MessagePublisher', 'EntitySchemaManager',
    'ProcessFlowElementSchema', 'SysSettings', 'GlobalSearch'
  ];

  private static readonly CATEGORIES = [
    'UI', 'Business Logic', 'Integration', 'Configuration', 'Data',
    'Security', 'Reporting', 'Workflow', 'API', 'System'
  ];

  /**
   * Generate a single video content object
   */
  static generateVideoContent(overrides: Partial<VideoContent> = {}): VideoContent {
    const id = faker.string.uuid();
    const topics = faker.helpers.arrayElements(this.TOPICS, { min: 2, max: 5 });
    const commands = faker.helpers.arrayElements(this.COMMANDS, { min: 1, max: 4 });
    const apiReferences = faker.helpers.arrayElements(this.API_REFERENCES, { min: 1, max: 3 });

    return {
      id,
      title: faker.lorem.words({ min: 3, max: 6 }),
      filePath: `/videos/${id}.mp4`,
      duration: faker.number.float({ min: 30, max: 1800 }),
      transcript: faker.lorem.paragraph({ min: 10, max: 20 }),
      summary: faker.lorem.paragraph({ min: 2, max: 4 }),
      topics,
      complexityLevel: faker.helpers.arrayElement(['beginner', 'intermediate', 'advanced']),
      commands,
      apiReferences,
      codeExamples: this.generateCodeExamples(),
      createdAt: faker.date.past({ years: 1 }),
      ...overrides,
    };
  }

  /**
   * Generate multiple video content objects
   */
  static generateVideoContents(count: number): VideoContent[] {
    return Array.from({ length: count }, () => this.generateVideoContent());
  }

  /**
   * Generate a single PDF content object
   */
  static generatePDFContent(overrides: Partial<PDFContent> = {}): PDFContent {
    const id = faker.string.uuid();
    const pageCount = faker.number.int({ min: 5, max: 50 });
    const topics = faker.helpers.arrayElements(this.TOPICS, { min: 2, max: 4 });

    return {
      id,
      title: faker.lorem.words({ min: 4, max: 8 }),
      filePath: `/pdfs/${id}.pdf`,
      pageCount,
      content: faker.lorem.paragraphs({ min: 10, max: 30 }),
      sections: this.generatePDFSections(pageCount),
      topics,
      commands: [],
      apiReferences: [],
      codeExamples: [],
      createdAt: faker.date.past({ years: 1 }),
      ...overrides,
    };
  }

  /**
   * Generate multiple PDF content objects
   */
  static generatePDFContents(count: number): PDFContent[] {
    return Array.from({ length: count }, () => this.generatePDFContent());
  }

  /**
   * Generate a single command object
   */
  static generateCommand(overrides: Partial<Command> = {}): Command {
    return {
      id: faker.number.int({ min: 1, max: 10000 }),
      command: faker.helpers.arrayElement(this.COMMANDS),
      description: faker.lorem.sentence(),
      category: faker.helpers.arrayElement(this.CATEGORIES),
      parameters: faker.helpers.arrayElements(
        ['name', 'value', 'type', 'required', 'default'],
        { min: 0, max: 3 }
      ),
      ...overrides,
    };
  }

  /**
   * Generate multiple command objects
   */
  static generateCommands(count: number): Command[] {
    return Array.from({ length: count }, () => this.generateCommand());
  }

  /**
   * Generate code examples
   */
  private static generateCodeExamples(): CodeExample[] {
    const languages = ['javascript', 'csharp', 'sql'];
    const selectedLanguages = faker.helpers.arrayElements(languages, { min: 1, max: 2 });
    
    return selectedLanguages.map(lang => ({
      language: lang,
      code: this.generateCodeForLanguage(lang),
      description: faker.lorem.sentence(),
    }));
  }

  /**
   * Generate code for specific language
   */
  private static generateCodeForLanguage(language: string): string {
    switch (language) {
      case 'javascript':
        return `var schema = this.Ext.create('EntitySchema', {
    name: '${faker.lorem.word()}',
    columns: {
        '${faker.lorem.word()}': {
            dataValueType: Terrasoft.DataValueType.TEXT
        }
    }
});`;
      case 'csharp':
        return `var entity = UserConnection.EntitySchemaManager.GetInstanceByName('${faker.lorem.word()}');
var query = new EntitySchemaQuery(entity);
query.AddAllSchemaColumns();
var result = query.GetEntityCollection(UserConnection);`;
      case 'sql':
        return `SELECT Id, Name, CreatedOn 
FROM ${faker.lorem.word()} 
WHERE CreatedOn > @StartDate 
ORDER BY CreatedOn DESC`;
      default:
        return faker.lorem.paragraph();
    }
  }

  /**
   * Generate PDF sections
   */
  private static generatePDFSections(pageCount: number): PDFSection[] {
    const sectionCount = faker.number.int({ min: 3, max: 8 });
    return Array.from({ length: sectionCount }, () => ({
      title: faker.lorem.words({ min: 2, max: 5 }),
      content: faker.lorem.paragraphs({ min: 3, max: 7 }),
      pageNumber: faker.number.int({ min: 1, max: pageCount }),
    }));
  }

  /**
   * Generate complete test dataset
   */
  static generateTestDataset(options: {
    videoCount?: number;
    pdfCount?: number;
    commandCount?: number;
  } = {}): {
    videos: VideoContent[];
    pdfs: PDFContent[];
    commands: Command[];
    metadata: {
      generatedAt: string;
      videoCount: number;
      pdfCount: number;
      commandCount: number;
    };
  } {
    const {
      videoCount = 50,
      pdfCount = 30,
      commandCount = 20,
    } = options;

    const videos = this.generateVideoContents(videoCount);
    const pdfs = this.generatePDFContents(pdfCount);
    const commands = this.generateCommands(commandCount);

    return {
      videos,
      pdfs,
      commands,
      metadata: {
        generatedAt: new Date().toISOString(),
        videoCount: videos.length,
        pdfCount: pdfs.length,
        commandCount: commands.length,
      },
    };
  }

  /**
   * Save test dataset to file
   */
  static async saveTestDataset(
    filePath: string,
    options: Parameters<typeof TestDataGenerator.generateTestDataset>[0] = {}
  ): Promise<void> {
    const dataset = this.generateTestDataset(options);
    const dir = path.dirname(filePath);
    
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(filePath, JSON.stringify(dataset, null, 2));
  }

  /**
   * Load test dataset from file
   */
  static async loadTestDataset(filePath: string): Promise<ReturnType<typeof TestDataGenerator.generateTestDataset>> {
    const content = await fs.readFile(filePath, 'utf-8');
    return JSON.parse(content);
  }

  /**
   * Generate minimal test data for unit tests
   */
  static generateMinimalTestData(): {
    video: VideoContent;
    pdf: PDFContent;
    command: Command;
  } {
    return {
      video: this.generateVideoContent(),
      pdf: this.generatePDFContent(),
      command: this.generateCommand(),
    };
  }
}

// CLI interface for generating test data
if (require.main === module) {
  const outputPath = path.join(__dirname, 'test-data.json');
  
  TestDataGenerator.saveTestDataset(outputPath, {
    videoCount: 100,
    pdfCount: 50,
    commandCount: 30,
  })
    .then(() => {
      console.log(`Test data generated and saved to ${outputPath}`);
    })
    .catch((error) => {
      console.error('Error generating test data:', error);
      process.exit(1);
    });
}
